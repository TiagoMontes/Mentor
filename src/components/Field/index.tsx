import {
  UseFormRegister,
  RegisterOptions,
  FieldValues,
  Path
} from 'react-hook-form'
import { twMerge } from 'tailwind-merge'

interface FieldProps<TFormValues extends FieldValues> {
  inputName: Path<TFormValues>
  register: UseFormRegister<TFormValues>
  type?: string
  rules?: RegisterOptions<TFormValues, Path<TFormValues>>
  className?: string
}

export default function Field<TFormValues extends FieldValues>({
  inputName,
  register,
  type = 'text',
  rules = {},
  className
}: FieldProps<TFormValues>) {
  return (
    <input
      className={twMerge(
        className,
        'w-full rounded-lg border-2 border-black p-4'
      )}
      {...register(inputName, rules)}
      type={type}
    />
  )
}
