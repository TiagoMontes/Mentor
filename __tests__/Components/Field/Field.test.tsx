import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import { RegisterOptions } from 'react-hook-form'
import Field from '@/components/Field'

type TestFormData = {
  name: string
  email: string
  age: number
}

const mockRegister = jest.fn().mockReturnValue({
  name: '',
  onChange: jest.fn(),
  onBlur: jest.fn(),
  ref: jest.fn(),
})

describe('Field Component', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders the input element with default props', () => {
    render(<Field<TestFormData> inputName="name" register={mockRegister} />)
    const input = screen.getByRole('textbox')
    expect(input).toBeInTheDocument()
    expect(input).toHaveClass('w-full rounded-lg border-2 border-black p-2')
    expect(input).toHaveAttribute('type', 'text')
  })

  it('applies the placeholder prop correctly', () => {
    render(
      <Field<TestFormData>
        inputName="name"
        register={mockRegister}
        placeholder="Digite seu nome"
      />
    )
    const input = screen.getByPlaceholderText('Digite seu nome')
    expect(input).toBeInTheDocument()
    expect(input).toHaveAttribute('placeholder', 'Digite seu nome')
  })

  it('merges custom className with default classes', () => {
    render(
      <Field<TestFormData>
        inputName="name"
        register={mockRegister}
        className="bg-blue-500 text-white"
      />
    )
    const input = screen.getByRole('textbox')
    expect(input).toHaveClass(
      'bg-blue-500 text-white w-full rounded-lg border-2 border-black p-2'
    )
  })

  it('applies the correct type prop', () => {
    render(
      <Field<TestFormData>
        inputName="email"
        register={mockRegister}
        type="email"
      />
    )
    const input = screen.getByRole('textbox')
    expect(input).toHaveAttribute('type', 'email')
  })

  it('registers the input with rules correctly', () => {
    const rules: RegisterOptions<TestFormData, 'name'> = { required: true }
    render(
      <Field<TestFormData>
        inputName="name"
        register={mockRegister}
        rules={rules}
      />
    )
    expect(mockRegister).toHaveBeenCalledWith('name', rules)
  })

  it('triggers onChange when typing', () => {
    render(<Field<TestFormData> inputName="name" register={mockRegister} />)
    const input = screen.getByRole('textbox')
    fireEvent.change(input, { target: { value: 'Teste' } })
    expect(mockRegister().onChange).toHaveBeenCalled()
  })

  it('renders without a label by default', () => {
    render(<Field<TestFormData> inputName="name" register={mockRegister} />)
    const label = screen.queryByRole('label')
    expect(label).not.toBeInTheDocument()
  })

  it('matches snapshot', () => {
    const { container } = render(
      <Field<TestFormData>
        inputName="name"
        register={mockRegister}
        placeholder="Digite seu nome"
        className="bg-red-500"
      />
    )
    expect(container).toMatchSnapshot()
  })
})
