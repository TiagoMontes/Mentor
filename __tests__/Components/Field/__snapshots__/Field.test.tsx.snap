// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Component matches snapshot 1`] = `
<div>
  <input
    class="bg-red-500 w-full rounded-lg border-2 border-black p-2"
    name=""
    placeholder="Digite seu nome"
    type="text"
  />
</div>
`;
